### DISCUSSION TITLE
Retrieve file path during S3TransferManager.uploadDirectory

### DISCUSSION THREAD

**Discussion by asemeria** [aws/aws-sdk-java-v2#4273] [Q&A]
I want to upload a directory with the new version of S3TransferManager v2 (https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/transfer-manager.html).

Everything works well except for the content type: I need to set the file content type according to the file name. For this reason, I thought to use the

`UploadDirectoryRequest.Builder uploadFileRequestTransformer(Consumer<UploadFileRequest.Builder> uploadFileRequestTransformer)`

But I don't know how can I retrieve the file path of the file that is going to be uploaded

Do you have any idea?

**Comment by sin0emp** [comment]
For any lost soul out there, I ended up doing this which is kind of stupid, I know! But I did not find any other way which was allowed by the SDK's API. Also, using `putObjectRequest` builder lambda directly, was throwing random null reference exceptions for me that's why I cloned another builder and set the content type like this.

```
var uploadRequest = UploadDirectoryRequest.builder()
                           .bucket(bucketName)
                           .s3Prefix(prefix)
                           .source(localPathObject)
                           .uploadFileRequestTransformer(builder -> {
                                   var built = builder.build();
                                   if (built.source().endsWith(".json")) {
                                           var newBuilder = built.putObjectRequest().toBuilder();
                                           newBuilder.contentType("application/json");
                                           builder.putObjectRequest(newBuilder.build());
                                   }
                           }).build();
```

**Comment by orispok** [comment]
To simplify, you can use `Files.probeContentType` to dynamically set the content type:

```
    val uploadDirectoryRequest = UploadDirectoryRequest.builder()
    .source(Paths.get(sourceDirectory))
    .bucket(bucketName)
    .uploadFileRequestTransformer { builder ->
        val built = builder.build();
        val newBuilder = built.putObjectRequest().toBuilder();
        val mimeType = Files.probeContentType(built.source())
        println("Mime type [$mimeType]. from file [${built.source()}].")
        newBuilder.contentType(mimeType);
        builder.putObjectRequest(newBuilder.build());
    }
    .build()
```
